typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1038 @@
+
+Then in `ArithExpr`, we simplify `Equality<A, B>` into an identity `(a: A) => A`, which gives us proof representation as a value.
+
+```typescript
+type ArithExpr<A> =
+  | { type: "Num"; v: number; proof: Equality<number, A> }
+  | { type: "Plus"; l: ArithExpr<number>; r: ArithExpr<number>; proof: Equality<number, A> }
+  | { type: "Gt"; l: ArithExpr<number>; r: ArithExpr<number>; proof: Equality<boolean, A> }
+  | { type: "And"; l: ArithExpr<boolean>; r: ArithExpr<boolean>; proof: Equality<boolean, A> };
+```
+
+Now let's define our helper functions which will help us to build expressions and use `identity<A>(a: A): A` as a proof.
+
+```typescript
+const num: (v: number) => ArithExpr<number> = (v) => {
+  return { type: "Num", v, proof: identity };
+};
+
+const plus: (l: ArithExpr<number>, r: ArithExpr<number>) => ArithExpr<number> = (l, r) => {
+  return { type: "Plus", l, r, proof: identity };
+};
+
+const gt: (l: ArithExpr<number>, r: ArithExpr<number>) => ArithExpr<boolean> = (l, r) => {
+  return { type: "Gt", l, r, proof: identity };
+};
+
+const and: (l: ArithExpr<boolean>, r: ArithExpr<boolean>) => ArithExpr<boolean> = (l, r) => {
+  return { type: "And", l, r, proof: identity };
+};
+```
+
+The last step will be just the implementation of `interpret`.
+With a switch statement on `expr.type`, we construct our result by running `interpret` on nested expressions and proving the resulting types with the proof from this `expr`.
+
+```typescript
+const interpret: <A>(expr: ArithExpr<A>) => A = (expr) => {
+  switch (expr.type) {
+    case "Num":
+      return expr.proof(expr.v);
+    case "Plus":
+      return expr.proof(interpret(expr.l) + interpret(expr.r));
+    case "Gt":
+      return expr.proof(interpret(expr.l) > interpret(expr.r));
+    case "And":
+      return expr.proof(interpret(expr.l) && interpret(expr.r));
+    default:
+      return expr;
+  }
+};
+
+const testExpr = and(gt(plus(num(23), num(12)), num(170)), gt(num(35), num(47)));
+interpret(testExpr); // false
+
+const wrongExpr = and(num(23), num(12)); // Argument of type 'ArithExpr<number>' is not assignable to parameter of type 'ArithExpr<boolean>'.
+```
+
+For a deeper understanding of this, you can check the [original](http://code.slipthrough.net/2016/08/10/approximating-gadts-in-purescript/) PureScript article and also Giulio Canti's [example](https://gist.github.com/gcanti/9a0c2a666621f03b80457831ff3ab997) (on which this part is based on) of its implementation in TypeScript.
+
+#### Tagless final eDSL
+
+Another way to implement an eDSL is [tagless final](https://serokell.io/blog/introduction-tagless-final).
+We can move from data type to type classes with the same logic.
+Here the type class will describe possible actions, and by using instances, we will be able to write interpreters.
+
+Let's start by defining our type class.
+To do this, we need the already-mentioned `URIS`(which is a union of all 1-arity defined in `URItoKind`) and the already-mentioned `Kind`.
```

Comment:
You can also say here, that you represent the type class as a dictionary of related methods carried in a mere type. That jump from Haskell type classes to the TS code with a record type was just too big.

Reply:
It was described in `HKT` part. But I think your explanation is better, so I can move it to the `HKT` part

Reply:
Ah, okay, I misunderstood you \

Reply:
Ah no, looks good to me now. Here you said "defining our type class" but TS has no typeclasses. And now when you in HKT section describe what you mean under TS typeclasses, "defining our type class" makes sense.